rir/src/compiler/util/phi_placement.cpp

```diff
@@ -0,0 +1,60 @@
+#include "phi_placement.h"
+#include "../pir/pir_impl.h"
+#include "../util/cfg.h"
+
+namespace rir {
+namespace pir {
+
+BB* PhiPlacement::find(const CFG& cfg, BB* searchBlock,
+                       const std::vector<BB*>& inputs) {
+    auto hasAllInputs = [&](BB* candidate) {
+        for (auto& input : inputs) {
+            // we cannot move the phi above its src
+            if (input == candidate)
+                return false;
+            if (!cfg.isPredecessor(input, candidate))
+                return false;
+        }
+        return true;
+    };
+    assert(hasAllInputs(searchBlock));
+
+    // First move up the phi until at least one input comes from a different
+    // block.
+    while (true) {
+        auto& preds = cfg.immediatePredecessors(searchBlock);
+        assert(!preds.empty());
+
+        for (auto pre : preds)
+            if (!hasAllInputs(pre))
+                goto checkCandidate;
+
+        searchBlock = *preds.begin();
```

Comment:
How does this work? Do we know here that preds has only one block?

Reply:
this is the case where all predecessors have all inputs. that means the graph looks like:
```
  a ... b
   \  /
  /  \
  \  /
   searchBlock
```
so it does not matter which pred we follow.